features/mbedtls/targets/TARGET_NUVOTON/TARGET_NUC472/sha/sha_alt_hw.c

```diff
@@ -260,71 +255,72 @@ void crypto_sha_update_nobuf(crypto_sha_context *ctx, const unsigned char *input
                 }
                 CRPT->SHA_DMACNT = lastblock_size;
                 CRPT->SHA_CTL = sha_ctl_start | CRPT_SHA_CTL_DMALAST_Msk;
-            }
-            else {
+            } else {
                 switch (sha_opmode) {
-                    case SHA_MODE_SHA256:
-                        DGST7_old = CRPT->SHA_DGST7;
-                    case SHA_MODE_SHA224:
-                        DGST5_old = CRPT->SHA_DGST5;
-                        DGST6_old = CRPT->SHA_DGST6;
-                    case SHA_MODE_SHA1:
-                        DGST0_old = CRPT->SHA_DGST0;
-                        DGST1_old = CRPT->SHA_DGST1;
-                        DGST2_old = CRPT->SHA_DGST2;
-                        DGST3_old = CRPT->SHA_DGST3;
-                        DGST4_old = CRPT->SHA_DGST4;
+                case SHA_MODE_SHA256:
+                    DGSTs[7] = CRPT->SHA_DGST7;
+                case SHA_MODE_SHA224:
+                    DGSTs[5] = CRPT->SHA_DGST5;
+                    DGSTs[6] = CRPT->SHA_DGST6;
+                case SHA_MODE_SHA1:
+                    DGSTs[0] = CRPT->SHA_DGST0;
+                    DGSTs[1] = CRPT->SHA_DGST1;
+                    DGSTs[2] = CRPT->SHA_DGST2;
+                    DGSTs[3] = CRPT->SHA_DGST3;
+                    DGSTs[4] = CRPT->SHA_DGST4;
                 }
 
                 CRPT->SHA_CTL = sha_ctl_start;
             }
-        }
-        else {  // Non-last word of a complete block
+        } else { // Non-last word of a complete block
             CRPT->SHA_CTL = sha_ctl_start;
         }
         while (! (CRPT->SHA_STS & CRPT_SHA_STS_DATINREQ_Msk));
         CRPT->SHA_DATIN = data;
-        
+
         in_pos += 4;
         rmn -= 4;
     }
-    
+
     if (islast) {   // Finish of last block
         while (CRPT->SHA_STS & CRPT_SHA_STS_BUSY_Msk);
-    }
-    else {  // Finish of non-last block
+    } else { // Finish of non-last block
         // No H/W flag to indicate finish of non-last block process.
         // Values of SHA_DGSTx registers will change as last word of the block is input, so use it for judgement.
         int isfinish = 0;
         while (! isfinish) {
             switch (sha_opmode) {
-                case SHA_MODE_SHA256:
-                    if (DGST7_old != CRPT->SHA_DGST7) {
-                        isfinish = 1;
-                        break;
-                    }
-                case SHA_MODE_SHA224:
-                    if (DGST5_old != CRPT->SHA_DGST5 || DGST6_old != CRPT->SHA_DGST6) {
-                        isfinish = 1;
-                        break;
-                    }
-                case SHA_MODE_SHA1:
-                    if (DGST0_old != CRPT->SHA_DGST0 || DGST1_old != CRPT->SHA_DGST1 || DGST2_old != CRPT->SHA_DGST2 ||
-                        DGST3_old != CRPT->SHA_DGST3 || DGST4_old != CRPT->SHA_DGST4) {
-                        isfinish = 1;
-                        break;
-                    }
+            case SHA_MODE_SHA256:
```

Comment:
@ccli8 Could you elaborate (here and perhaps in a comment) why we need this loop? One reason I can conceive is that if we want to clone a H/W context, we should be sure that it's not in the middle of its internal operations. Are there more reasons? (I'm not arguing the loop shouldn't be there, but I'd like to understand the original motivation for introducing it). 

It would be very unfortunate if there was a fixed point for the internal operation - in this case we'd loop indefinitely.

Reply:
@hanno-arm I refine the comment in https://github.com/ARMmbed/mbed-os/pull/4925/commits/82a224d5703be289e104ddfcb09cfde4d1837bec. There's no internal operation on-going after exiting `crypto_sha_update_nobuf`, so H/W context can be safely cloned to S/W context out of `crypto_sha_update_nobuf`.

For loop indefinitely, I add timeout mechanism in the same commit to detect two SHA digest code are the same in a row.

Reply:
@ccli8 Thank you for enhancing the code. So there's no other reason than cloning why we'd want the internal operations to be complete once `crypto_sha_update_nobuf` returns?

Reply:
@hanno-arm `crypto_sha_update_nobuf`'s current behavior is blocking, meaning digest code is ready after its return. With that guarantee, we can call `crypto_sha_getinternstate` to get intermediary/final digest code from e.g. `mbedtls_sha1_clone`/`mbedtls_sha1_finish` safely. That's why there are `while` loops in the end to wait for its finish (by block or total).